controllers/baseboardmanagement_controller.go

```diff
@@ -18,42 +18,233 @@ package controllers
-// the BaseboardManagement object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
-//
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.11.0/pkg/reconcile
+// Reconcile ensures the state of a BaseboardManagement.
+// Gets the BaseboardManagement object and uses the SecretReference to initialize a BMC Client.
+// Ensures the BMC power is set to the desired state.
+// Updates the status and conditions accordingly.
 func (r *BaseboardManagementReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := r.logger.WithValues("BaseboardManagement", req.NamespacedName)
+	logger.Info("Reconciling BaseboardManagement")
+
+	// Fetch the BaseboardManagement object
+	baseboardManagement := &bmcv1alpha1.BaseboardManagement{}
+	err := r.client.Get(ctx, req.NamespacedName, baseboardManagement)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+
+		logger.Error(err, "Failed to get BaseboardManagement")
+		return ctrl.Result{}, err
+	}
+
+	// Deletion is a noop.
+	if !baseboardManagement.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, nil
+	}
+
+	return r.reconcile(ctx, baseboardManagement, logger)
+}
+
+func (r *BaseboardManagementReconciler) reconcile(ctx context.Context, bm *bmcv1alpha1.BaseboardManagement, logger logr.Logger) (ctrl.Result, error) {
+	// Fetching username, password from SecretReference
+	// Requeue if error fetching secret
+	username, password, err := r.resolveAuthSecretRef(ctx, bm.Spec.Connection.AuthSecretRef)
+	if err != nil {
+		return ctrl.Result{Requeue: true}, fmt.Errorf("resolving BaseboardManagement %s/%s SecretReference: %v", bm.Namespace, bm.Name, err)
+	}
+
+	// TODO (pokearu): Remove port hardcoding
+	// Initializing BMC Client
+	bmcClient, err := r.bmcClientFactory(ctx, bm.Spec.Connection.Host, "623", username, password)
+	if err != nil {
+		logger.Error(err, "BMC connection failed", "host", bm.Spec.Connection.Host)
+		result, setConditionErr := r.setCondition(ctx, bm, bmcv1alpha1.ConnectionError, err.Error())
+		if setConditionErr != nil {
+			return result, utilerrors.NewAggregate([]error{fmt.Errorf("failed to set conditions: %v", setConditionErr), err})
+		}
+		return result, err
+	}
+
+	// Close BMC connection after reconcilation
+	defer func() {
+		err = bmcClient.Close(ctx)
+		if err != nil {
+			logger.Error(err, "BMC close connection failed", "host", bm.Spec.Connection.Host)
+		}
+	}()
+
+	// fieldReconcilers defines BaseboardManagement spec field reconciler functions
+	fieldReconcilers := []baseboardManagementFieldReconciler{
+		r.reconcilePower,
+	}
+	for _, reconiler := range fieldReconcilers {
+		if err := reconiler(ctx, bm, bmcClient); err != nil {
+			logger.Error(err, "Failed to reconcile BaseboardManagement", "host", bm.Spec.Connection.Host)
```

Comment:
I think you still want to update the status on failure. You want to converge on desired state, not make it all-or-nothing

Reply:
As discussed, I modified the reconciler to update status when reconciling fields 👍